sbndaq-artdaq/ArtModules/SBND/SBNDGateFilter_module.cc

```diff
@@ -0,0 +1,205 @@
+
+#include "sbndaq-artdaq-core/Overlays/SBND/PTBFragment.hh"
+#include "artdaq-core/Data/ContainerFragment.hh"
+
+
+#include "artdaq/DAQdata/Globals.hh"
+#include "artdaq-core/Data/Fragment.hh"
+
+#include <algorithm>
+#include <cassert>
+#include <cmath>
+#include <fstream>
+#include <iomanip>
+#include <vector>
+#include <iostream>
+
+#ifdef TRACE_NAME
+#undef TRACE_NAME
+#endif
+
+#define TRACE_NAME "SBNDGateFilter_module"
+
+namespace sbnd {
+  class SBNDGateFilter;
+}
+
+
+class sbnd::SBNDGateFilter : public art::EDFilter {
+public:
+  explicit SBNDGateFilter(fhicl::ParameterSet const & pset);
+  virtual ~SBNDGateFilter();
+
+  virtual bool filter(art::Event & evt) override;
+  //template <class T>  
+  std::vector<uint64_t> GetAllHLTs(artdaq::ContainerFragment *trigfrag);
+  std::vector<uint64_t> GetHLT(sbndaq::CTBFragment ptb_fragment);
+  bool ApplyGateFilter(std::vector<uint64_t> triggers); //(artdaq::ContainerFragment *trigfrag);
+
+private:
+  std::string fInputLabel, fInputContainerInstance, fInputNonContainerInstance;
+  std::vector<int> ftrigger_type;
+  std::vector<uint64_t> fexcluded_triggers;
+};
+
+sbnd::SBNDGateFilter::SBNDGateFilter(fhicl::ParameterSet const & pset)
+  : EDFilter(pset),
+    fInputLabel(pset.get<std::string>("InputLabel", "daq")),
+    fInputContainerInstance(pset.get<std::string>("InputContainerInstance", "ContainerPTB")), 
+    fInputNonContainerInstance(pset.get<std::string>("InputNonContainerInstance", "PTB")), 
+    ftrigger_type(pset.get<std::vector<int>>("trigger_type",{-1})),
+    fexcluded_triggers( pset.get< std::vector<uint64_t> >("excluded_triggers", {}))
+{
+}
+
+sbnd::SBNDGateFilter::~SBNDGateFilter()
+{
+}
+
+std::vector<uint64_t> sbnd::SBNDGateFilter::GetAllHLTs(artdaq::ContainerFragment*  ptb_container_fragment){
+  std::vector<uint64_t> triggers;  
+  
+  for (size_t f=0; f<ptb_container_fragment->block_count(); ++f){//loop over container of fragments
+    artdaq::Fragment frag=*ptb_container_fragment->at(f).get();
+    sbndaq::CTBFragment ptb_fragment(frag);
+    //==============
+    std::vector fragTriggers=GetHLT(ptb_fragment); //Not sure that there could really be multiple HLTs in a single fragment but just in case I'll make it vector
+    triggers.insert(triggers.end(), fragTriggers.begin(), fragTriggers.end() );//append list of triggers in this fragment to all of the HLTs in the container
+  }//end loop over fragments
+  
+  return triggers;
+}
+
+
+
+std::vector<uint64_t> sbnd::SBNDGateFilter::GetHLT(sbndaq::CTBFragment ptb_fragment){
+  std::vector<uint64_t> triggers;  
+
+  for ( size_t i = 0; i < ptb_fragment.NWords(); i++ ) {//loop over words in fragment       
+    //if  (ptb_fragment.Word(i)->IsHLT()==false) continue;  
+    if  (ptb_fragment.Word(i)->word_type !=0x2 ) continue; //0x2 is the type for an HLT (0x1 for LLT) 
+    uint64_t hlttrigger=ptb_fragment.Trigger(i)->trigger_word & 0x1FFFFFFFFFFFFFFF;
+    hlttrigger= log(1.*hlttrigger)/log(2.) ; 
```

Comment:
This line will give a wrong number in the edge case when 2 HLTs appeared at the same time. This can happen for example if there is light threshold crossed at the same time as we open the beam gate and that gate was about to make zero bias trigger.

Here's a suggestion how to adjust lines 93-97:
``` 
 uint64_t hlt_mask = ptb_fragment.Trigger(i)->trigger_word & 0x1FFFFFFFFFFFFFFF;
 // Process each set bit in hlt_mask as a separate HLT trigger
        while (hlt_mask) {
            uint64_t hlttrigger = __builtin_ctzll(hlt_mask); // Find the least significant set bit
            hlt_mask &= (hlt_mask - 1); // Clear the least significant set bit
            if (hlttrigger >= 20) continue;  //HLT triggers greater then 20 are reserved for non event triggers
            triggers.emplace_back(hlttrigger);
        }
```